hawc/apps/epi/api.py

```diff
@@ -87,37 +100,425 @@ def result_heatmap(self, request, pk):
         return Response(export)
 
 
-class StudyPopulation(AssessmentViewset):
+class Criteria(PermCheckerMixin, AssessmentEditViewset):
+    perm_checker_key = "assessment"
+    assessment_filter_args = "assessment"
+    model = models.Criteria
+    serializer_class = serializers.CriteriaSerializer
+
+
+class StudyPopulation(PermCheckerMixin, viewsets.ModelViewSet):
+    perm_checker_key = "study"
     assessment_filter_args = "study__assessment"
     model = models.StudyPopulation
     serializer_class = serializers.StudyPopulationSerializer
 
+    criteria_categories = (
+        ("inclusion_criteria", "I",),
+        ("exclusion_criteria", "E",),
+        ("confounding_criteria", "C",),
+    )
+
+    def process_criteria_association(self, serializer, study_population_id, post_initial_create):
+        # this should probably be rewritten to use the ManyToManyManager on the underlying instance...
+        inserts = []
+        for cc in self.criteria_categories:
+            data_key = cc[0]
+            type_code = cc[1]
+
+            if data_key in self.request.data:
+                if not post_initial_create:
+                    # wipe out existing criteria for this pop+type pair that was part of the request...
+                    models.StudyPopulationCriteria.objects.filter(
+                        study_population=self.get_object(), criteria_type=type_code
+                    ).delete()
+
+                criteria_ids = self.request.data[data_key]
+
+                for criteria_id in criteria_ids:
+                    dynamic_obj = {
+                        "criteria_type": type_code,
+                        "criteria": criteria_id,
+                        "study_population": study_population_id,
+                    }
+                    inserts.append(dynamic_obj)
+
+        # ...and save any new ones
+        if len(inserts) > 0:
+            serializer = serializers.SimpleStudyPopulationCriteriaSerializer(
+                data=inserts, many=True
+            )
+            serializer.is_valid(raise_exception=True)
+            serializer.save()
+
+    def get_queryset(self, *args, **kwargs):
+        return self.model.objects.all()
+
+    def handle_criteria(self, request, during_create):
```

Comment:
i think the serializer should have a study instance you can just use. `serializer.instance.study`; no need for more db calls